wdl-engine/src/backend/crankshaft.rs

```diff
@@ -397,41 +399,79 @@ impl TaskExecutionBackend for CrankshaftBackend {
         Some(Path::new(GUEST_WORK_DIR))
     }
 
-    fn localize_inputs<'a, 'b, 'c, 'd>(
+    fn localize_inputs<'a, 'b, 'c>(
         &'a self,
-        downloader: &'b dyn Downloader,
+        downloader: Arc<dyn Downloader>,
         inputs: &'c mut [crate::eval::Input],
-    ) -> BoxFuture<'d, Result<()>>
+    ) -> BoxFuture<'c, Result<()>>
     where
-        'a: 'd,
-        'b: 'd,
-        'c: 'd,
-        Self: 'd,
+        'a: 'c,
+        'b: 'c,
+        Self: 'c,
     {
-        async {
+        async move {
             // Construct a trie for mapping input guest paths
             let mut trie = InputTrie::default();
             for input in inputs.iter() {
                 trie.insert(input)?;
             }
 
             for (index, guest_path) in trie.calculate_guest_paths(GUEST_INPUTS_DIR)? {
-                inputs[index].set_guest_path(guest_path);
+                if let Some(input) = inputs.get_mut(index) {
+                    input.set_guest_path(guest_path);
+                } else {
+                    return Err(anyhow!("invalid index {} returned from trie", index));
+                }
             }
 
             // Localize all inputs
             // TODO: only do this for local task execution
-            for input in inputs {
-                // TODO: parallelize the downloads
-                let location = match input.path() {
-                    EvaluationPath::Local(path) => Location::Path(path.into()),
-                    EvaluationPath::Remote(url) => downloader
-                        .download(url)
-                        .await
-                        .map_err(|e| anyhow!("failed to localize `{url}`: {e:?}"))?,
-                };
+            let mut download_futs = JoinSet::new();
+
+            let urls_to_download: Vec<(usize, Url)> = inputs
```

Comment:
Here, I would recommend just iterating through the inputs mutably once and matching on the `EvaluationPath`:

- If it's an `EvaluationPath::Remote`, you can spawn the task to your `JoinSet` and collect the result later.
- If it's an `EvaluationPath::Local`, you set the location in place.

Upon completion of the first pass through, you can then do what you're doing in line 258. In other words, I these first three sections can be condensed down into one pass.